[lit] Add support to print paths relative to CWD in the test summary report

bjope opened the pull request:
This patch adds a -r|--relative-paths option to llvm-lit, which when enabled will print test case names using paths relative to the current working directory. The legacy default without that option is that test cases are identified using a path relative to the test suite.

Only the summary report is impacted. That normally include failed tests, unless unless options such as --show-pass.

Background to this patch was the discussion here
  https://discourse.llvm.org/t/test-case-paths-in-llvm-lit-output-are-lacking-the-location-of-the-test-dir-itself/87973
with a goal to making it easier to copy-n-paste the path to the failing test cases.

Examples showing difference in "Passed Tests" and "Failed Tests":

 > llvm-lit --show-pass test/Transforms/Foo
 PASS: LLVM :: Transforms/Foo/test1.txt (1 of 2)
 FAIL: LLVM :: Transforms/Foo/test2.txt (2 of 2)
 Passed Tests (1):
   LLVM :: Transforms/Foo/test1.txt
 Failed Tests (1):
   LLVM :: Transforms/Foo/test2.txt

 > llvm-lit --show-pass --relative-paths test/Transforms/Foo
 PASS: LLVM :: Transforms/Foo/test1.txt (1 of 2)
 FAIL: LLVM :: Transforms/Foo/test2.txt (2 of 2)
 Passed Tests (1):
   test/Transforms/Foo/test1.txt
 Failed Tests (1):
   test/Transforms/Foo/test2.txt

bjope committed:
[lit] Add support to print relative paths when outputting test names

This patch adds a -r|--relative-paths option to llvm-lit, which when
enabled will print test case names using paths relative to the
current working directory. The legacy default without that option
is that test cases are identified using a path relative to the
test suite.

Examples:
  > llvm-lit test/Transforms/ADCE/willreturn.ll
  PASS: LLVM :: Transforms/ADCE/willreturn.ll

  > llvm-lit -r test/Transforms/ADCE/willreturn.ll
  PASS: LLVM :: test/Transforms/ADCE/willreturn.ll

The option is currently only impacting how tests are identified in
console output. Not other kinds of reports.

Background to this patch was the discussion here:
  https://discourse.llvm.org/t/test-case-paths-in-llvm-lit-output-are-lacking-the-location-of-the-test-dir-itself/87973

nikic reviewed (commented):
Any reason not to enable this by default?

I guess one unfortunate interaction here would be with `ninja -Cbuild check-llvm`, which would presumably print tests relative to `build` instead of the root directory?

bjope committed:
Fixup formatting

bjope commented on llvm/utils/lit/lit/Test.py:
```diff
@@ -307,6 +307,11 @@ def isFailure(self):
 
+    def getFullNameRelPath(self, printRelativePaths):
+        if printRelativePaths:
+            return self.suite.config.name + " :: " + os.path.relpath(self.getFilePath())
```
In the summary report with failed tests (or when using `--show-pass` etc.) we can skip the test suite name and `::`.

My initial commit in this pull requst also change the output that list progress of tests such as 
`  PASS: LLVM :: Analysis/DemandedBits/basic.ll (1425 of 1444) `
but I think it would be enough to print the path relative cwd in the summary report.

Maybe I should limit this new option to those summary printouts (undoing the changes in `display.py`)?
(I'll prepare such an update...)

bjope committed:
Only impact summary report, skip "suite :: " prefix

Fixup to only impact paths in the summary report (undoing changes
in display.py).

Also skipping the "suite :: " prefix in such output.

bjope commented on llvm/utils/lit/tests/print-relative-path.py:
```diff
@@ -0,0 +1,18 @@
+# CHECK-RELATIVE: PASS: print-relative-path :: test.txt (1 of 2)
+# CHECK-RELATIVE: FAIL: print-relative-path :: test2.txt (2 of 2)
+# CHECK-RELATIVE: Passed Tests (1):
+# CHECK-RELATIVE:  Inputs/print-relative-path/test.txt
```
Need to use [/\\] to support windows paths. 
Preparing a fixup!
